mijia/src/bluetooth/events.rs

```diff
@@ -57,96 +61,144 @@ pub enum CharacteristicEvent {
 }
 
 impl BluetoothEvent {
-    /// Return a `MatchRule` which will match all D-Bus messages which represent Bluetooth events.
-    pub(crate) fn match_rule() -> MatchRule<'static> {
+    /// Return a set of `MatchRule`s which will match all D-Bus messages which represent Bluetooth
+    /// events, possibly limited to those for a particular object (such as a device, service or
+    /// characteristic).
+    ///
+    /// Note that the match rules for a device will not match the device discovered event for that
+    /// device, as it is considered an event for the system rather than the device itself.
+    pub(crate) fn match_rules(object: Option<impl Into<Path<'static>>>) -> Vec<MatchRule<'static>> {
         // BusName validation just checks that the length and format is valid, so it should never
         // fail for a constant that we know is valid.
         let bus_name = "org.bluez".into();
-        PropertiesPropertiesChanged::match_rule(Some(&bus_name), None).static_clone()
+
+        let mut match_rules = vec![];
+
+        // If we aren't filtering to a single device or characteristic, then match ObjectManager
+        // signals so we can get events for new devices being discovered.
+        if object.is_none() {
+            let match_rule =
+                ObjectManagerInterfacesAdded::match_rule(Some(&bus_name), None).static_clone();
+            match_rules.push(match_rule);
+        }
+
+        // Match PropertiesChanged signals for the given device or characteristic and all objects
+        // under it. If no object is specified then this will match PropertiesChanged signals for
+        // all BlueZ objects.
+        {
+            let object_path = object.map(|o| o.into());
+            let mut match_rule =
+                PropertiesPropertiesChanged::match_rule(Some(&bus_name), object_path.as_ref())
+                    .static_clone();
+            match_rule.path_is_namespace = true;
+            match_rules.push(match_rule);
+        }
```

Comment:
scope feels unnecessary

Reply:
Removed. I think I added because you made a comment in a previous code review about liking scopes for something similar to this.

Reply:
That was mostly me desperately trying to find a way to make sense of a 200-line test function. Splitting into separate functions would always be preferable for real code.